ringpop.thrift-gen

```diff
@@ -0,0 +1,171 @@
+    router      ringpop.Router
+}
+
+type {{ $ConfigName }} struct {
+    // structs for all service endpoints containing forwarding behavior
+    {{ range .Methods }}
+        {{ .Name | goPublicName }} *{{ $Service.Name }}{{ .Name | goPublicName }}Configuration
+    {{ end }}
+}
+// TODO make validation function on the config
+
+type tchan{{ $Service.Name }}Adaptable interface {
+    SetAdapter({{ $TChanInterface }})
+}
+
+func New{{ $Adapter }}(
+    impl   {{ $TChanInterface }},
+    rp     ringpop.Interface,
+    ch     *tchannel.Channel,
+    config {{ $ConfigName }},
+) *{{ $Adapter }} {
+    adapter := &{{ $Adapter }}{
+        impl:    impl,
+        ringpop: rp,
+        ch:      ch,
+        config:  config,
+    }
+    // create ringpop router for routing based on ring membership
+    adapter.router = ringpop.NewRouter(rp, adapter)
+
+    // notify implementation it is adapted
+    if adaptable, ok := impl.(tchan{{ $Service.Name }}Adaptable); ok {
+        adaptable.SetAdapter(adapter)
+    }
+
+    return adapter
+}
+
+// implement ringpop.ClientFactory
+func (a *{{ $Adapter }}) MakeClient(dest string) interface{} {
+    if dest == a.ringpop.WhoAmI() {
+        // use local client
+        return a.impl
+    }
+
+    // create remote client
+    client := thrift.NewClient(
+        a.ch,
+        a.ch.ServiceName(),
+        &thrift.ClientOptions{
+            HostPort: dest,
+        },
+    )
+    return New{{ $TChanInterface }}Client(client)
+}
+
+{{ range .Methods }}
+{{ $Method := .}}
+{{ $PublicName := $Method.Name | goPublicName }}
+{{ $cfg := print "config." $PublicName }}
+{{ $ConfigType := print ($Service.Name | goPublicName) $PublicName "Configuration"}}
+
+// ------ 'Generated' code for {{ $Method.Name | goPublicName }} ------ {{ $cfg }}
+type {{ $ConfigType }} struct {
+    N       ringpop.ReplicationFactor
+    Fanout  ringpop.ReplicationFactor
+    Reducer func({{ if $Method.ReturnType }}rs []{{ $Method.ReturnType | goType }}, {{ end }}errs []error) ({{ template "returnType" $Method}})
+    Key     func({{ template "arguments" $Method}}) (string, error)
+}
+
+func (a *{{ $Adapter }}) {{ $PublicName }} ({{ template "arguments" $Method }}) ({{ template "returnType" $Method }}) {
+    // check if the function is configured to be ring aware
+    if a.{{ $cfg }} == nil {
+        return a.impl.{{ $PublicName }}({{ template "call" $Method }})
+    }
+
+    if hasForwardedHeader(ctx) {
+        return a.impl.{{ $PublicName }}({{ template "call" $Method }})
+    }
+
+    // find the key to shard on
+    ringpopKey, err := a.{{ $cfg }}.Key({{ template "call" $Method }})
+    if err != nil {
+        return {{ if $Method.ReturnType }}r, {{ end }}fmt.Errorf("could not get key: %s", err)
+    }
+
+    clients := a.router.GetClients(ringpopKey, a.{{ $cfg }}.N, a.{{ $cfg }}.Fanout)
+    clientLength := len(clients)
+    if clientLength == 0 {
+        return {{ if $Method.ReturnType }}r, {{ end }}errors.New("no clients")
```

Comment:
optionally: for this error and the earlier error, it might be worth adding hte method name which returned that error. Otherwise, the remote side is going to get a very cryptic error "no clients". Try to add more information where possible to the errors, since they are the message the remote side will see.